File: gerrit-server/src/test/java/com/google/gerrit/server/ChangeUtilTest.java
// Copyright (C) 2017 The Android Open Source Project
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package com.google.gerrit.server;

import static com.google.common.truth.Truth.assertThat;

import com.google.gerrit.reviewdb.server.AccountAccess;
import com.google.gerrit.reviewdb.server.AccountExternalIdAccess;
import com.google.gerrit.reviewdb.server.AccountGroupAccess;
import com.google.gerrit.reviewdb.server.AccountGroupByIdAccess;
import com.google.gerrit.reviewdb.server.AccountGroupByIdAudAccess;
import com.google.gerrit.reviewdb.server.AccountGroupMemberAccess;
import com.google.gerrit.reviewdb.server.AccountGroupMemberAuditAccess;
import com.google.gerrit.reviewdb.server.AccountGroupNameAccess;
import com.google.gerrit.reviewdb.server.ChangeAccess;
import com.google.gerrit.reviewdb.server.ChangeMessageAccess;
import com.google.gerrit.reviewdb.server.PatchLineCommentAccess;
import com.google.gerrit.reviewdb.server.PatchSetAccess;
import com.google.gerrit.reviewdb.server.PatchSetApprovalAccess;
import com.google.gerrit.reviewdb.server.ReviewDb;
import com.google.gerrit.reviewdb.server.SchemaVersionAccess;
import com.google.gerrit.reviewdb.server.SystemConfigAccess;
import com.google.gwtorm.server.Access;
import com.google.gwtorm.server.StatementExecutor;

import org.junit.Test;

import java.util.concurrent.atomic.AtomicInteger;
import java.util.regex.Pattern;

public class ChangeUtilTest {
  @Test
  public void changeMessageUuid() throws Exception {
    Pattern pat = Pattern.compile("^[0-9a-f]{8}_[0-9a-f]{8}$");
    assertThat("abcd1234_0987fedc").matches(pat);

    ReviewDb db = new FakeDb();
    String id1 = ChangeUtil.messageUUID(db);
    assertThat(id1).matches(pat);

    String id2 = ChangeUtil.messageUUID(db);
    assertThat(id2).isNotEqualTo(id1);
    assertThat(id2).matches(pat);
  }

  private static class FakeDb implements ReviewDb {
    private final AtomicInteger SEQ = new AtomicInteger(100);

    @Override
    public int nextChangeMessageId() {
      return SEQ.getAndIncrement();
    }

    @Override
    public void commit() {
      throw new UnsupportedOperationException();
    }

    @Override
    public void rollback() {
      throw new UnsupportedOperationException();
    }

    @Override
    public void updateSchema(StatementExecutor e) {
      throw new UnsupportedOperationException();
    }

    @Override
    public void pruneSchema(StatementExecutor e) {
      throw new UnsupportedOperationException();
    }

    @Override
    public Access<?, ?>[] allRelations() {
      throw new UnsupportedOperationException();
    }

    @Override
    public void close() {
      throw new UnsupportedOperationException();
    }

    @Override
    public SchemaVersionAccess schemaVersion() {
      throw new UnsupportedOperationException();
    }

    @Override
    public SystemConfigAccess systemConfig() {
      throw new UnsupportedOperationException();
    }

    @Override
    public AccountAccess accounts() {
      throw new UnsupportedOperationException();
    }

    @Override
    public AccountExternalIdAccess accountExternalIds() {
      throw new UnsupportedOperationException();
    }

    @Override
    public AccountGroupAccess accountGroups() {
      throw new UnsupportedOperationException();
    }

    @Override
    public AccountGroupNameAccess accountGroupNames() {
      throw new UnsupportedOperationException();
    }

    @Override
    public AccountGroupMemberAccess accountGroupMembers() {
      throw new UnsupportedOperationException();
    }

    @Override
    public AccountGroupMemberAuditAccess accountGroupMembersAudit() {
      throw new UnsupportedOperationException();
    }

    @Override
    public ChangeAccess changes() {
      throw new UnsupportedOperationException();
    }

    @Override
    public PatchSetApprovalAccess patchSetApprovals() {
      throw new UnsupportedOperationException();
    }

    @Override
    public ChangeMessageAccess changeMessages() {
      throw new UnsupportedOperationException();
    }

    @Override
    public PatchSetAccess patchSets() {
      throw new UnsupportedOperationException();
    }

    @Override
    public PatchLineCommentAccess patchComments() {
      throw new UnsupportedOperationException();
    }

    @Override
    public AccountGroupByIdAccess accountGroupById() {
      throw new UnsupportedOperationException();
    }

    @Override
    public AccountGroupByIdAudAccess accountGroupByIdAud() {
      throw new UnsupportedOperationException();
    }

    @Override
    public int nextAccountId() {
      throw new UnsupportedOperationException();
    }

    @Override
    public int nextAccountGroupId() {
      throw new UnsupportedOperationException();
    }

    @Override
    public int nextChangeId() {
      throw new UnsupportedOperationException();
    }
  }
}

File: polygerrit-ui/app/utils/change-util_test.js
/**
 * @license
 * Copyright (C) 2017 The Android Open Source Project
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import '../test/common-test-setup-karma.js';
import {
  changeBaseURL,
  changePath,
  changeStatuses,
  changeStatusString,
  isRemovableReviewer,
} from './change-util.js';

suite('change-util tests', () => {
  let originalCanonicalPath;

  suiteSetup(() => {
    originalCanonicalPath = window.CANONICAL_PATH;
    window.CANONICAL_PATH = '/r';
  });

  suiteTeardown(() => {
    window.CANONICAL_PATH = originalCanonicalPath;
  });

  test('changeBaseURL', () => {
    assert.deepEqual(
        changeBaseURL('test/project', '1', '2'),
        '/r/changes/test%2Fproject~1/revisions/2'
    );
  });

  test('changePath', () => {
    assert.deepEqual(changePath('1'), '/r/c/1');
  });

  test('Open status', () => {
    const change = {
      change_id: 'Iad9dc96274af6946f3632be53b106ef80f7ba6ca',
      revisions: {
        rev1: {_number: 1},
      },
      current_revision: 'rev1',
      status: 'NEW',
      labels: {},
      mergeable: true,
    };
    let statuses = changeStatuses(change);
    const statusString = changeStatusString(change);
    assert.deepEqual(statuses, []);
    assert.equal(statusString, '');

    change.submittable = false;
    statuses = changeStatuses(change,
        {includeDerived: true});
    assert.deepEqual(statuses, ['Active']);

    // With no missing labels but no submitEnabled option.
    change.submittable = true;
    statuses = changeStatuses(change,
        {includeDerived: true});
    assert.deepEqual(statuses, ['Active']);

    // Without missing labels and enabled submit
    statuses = changeStatuses(change,
        {includeDerived: true, submitEnabled: true});
    assert.deepEqual(statuses, ['Ready to submit']);

    change.mergeable = false;
    change.submittable = true;
    statuses = changeStatuses(change,
        {includeDerived: true});
    assert.deepEqual(statuses, ['Merge Conflict']);

    delete change.mergeable;
    change.submittable = true;
    statuses = changeStatuses(change,
        {includeDerived: true, mergeable: true, submitEnabled: true});
    assert.deepEqual(statuses, ['Ready to submit']);

    change.submittable = true;
    statuses = changeStatuses(change,
        {includeDerived: true, mergeable: false});
    assert.deepEqual(statuses, ['Merge Conflict']);
  });

  test('Merge conflict', () => {
    const change = {
      change_id: 'Iad9dc96274af6946f3632be53b106ef80f7ba6ca',
      revisions: {
        rev1: {_number: 1},
      },
      current_revision: 'rev1',
      status: 'NEW',
      labels: {},
      mergeable: false,
    };
    const statuses = changeStatuses(change);
    const statusString = changeStatusString(change);
    assert.deepEqual(statuses, ['Merge Conflict']);
    assert.equal(statusString, 'Merge Conflict');
  });

  test('mergeable prop undefined', () => {
    const change = {
      change_id: 'Iad9dc96274af6946f3632be53b106ef80f7ba6ca',
      revisions: {
        rev1: {_number: 1},
      },
      current_revision: 'rev1',
      status: 'NEW',
      labels: {},
    };
    const statuses = changeStatuses(change);
    const statusString = changeStatusString(change);
    assert.deepEqual(statuses, []);
    assert.equal(statusString, '');
  });

  test('Merged status', () => {
    const change = {
      change_id: 'Iad9dc96274af6946f3632be53b106ef80f7ba6ca',
      revisions: {
        rev1: {_number: 1},
      },
      current_revision: 'rev1',
      status: 'MERGED',
      labels: {},
    };
    const statuses = changeStatuses(change);
    const statusString = changeStatusString(change);
    assert.deepEqual(statuses, ['Merged']);
    assert.equal(statusString, 'Merged');
  });

  test('Abandoned status', () => {
    const change = {
      change_id: 'Iad9dc96274af6946f3632be53b106ef80f7ba6ca',
      revisions: {
        rev1: {_number: 1},
      },
      current_revision: 'rev1',
      status: 'ABANDONED',
      labels: {},
    };
    const statuses = changeStatuses(change);
    const statusString = changeStatusString(change);
    assert.deepEqual(statuses, ['Abandoned']);
    assert.equal(statusString, 'Abandoned');
  });

  test('Open status with private and wip', () => {
    const change = {
      change_id: 'Iad9dc96274af6946f3632be53b106ef80f7ba6ca',
      revisions: {
        rev1: {_number: 1},
      },
      current_revision: 'rev1',
      status: 'NEW',
      is_private: true,
      work_in_progress: true,
      labels: {},
      mergeable: true,
    };
    const statuses = changeStatuses(change);
    const statusString = changeStatusString(change);
    assert.deepEqual(statuses, ['WIP', 'Private']);
    assert.equal(statusString, 'WIP, Private');
  });

  test('Merge conflict with private and wip', () => {
    const change = {
      change_id: 'Iad9dc96274af6946f3632be53b106ef80f7ba6ca',
      revisions: {
        rev1: {_number: 1},
      },
      current_revision: 'rev1',
      status: 'NEW',
      is_private: true,
      work_in_progress: true,
      labels: {},
      mergeable: false,
    };
    const statuses = changeStatuses(change);
    const statusString = changeStatusString(change);
    assert.deepEqual(statuses, ['Merge Conflict', 'WIP', 'Private']);
    assert.equal(statusString, 'Merge Conflict, WIP, Private');
  });

  test('isRemovableReviewer', () => {
    let change = {
      removable_reviewers: [{_account_id: 1}],
    };
    const reviewer = {_account_id: 1};

    assert.equal(isRemovableReviewer(change, reviewer), true);

    change = {
      removable_reviewers: [{_account_id: 2}],
    };
    assert.equal(isRemovableReviewer(change, reviewer), false);
  });
});

